### ISSUE TITLE
False positive for numbers without units in `declaration-property-value-no-unknown`

### ISSUE THREAD

**Issue by pamelalozano16** [stylelint/stylelint#8096]
### What minimal example or steps are needed to reproduce the bug?

```css
.example {
  left: 10; /* This should not be incompatible */
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgGxgDMd4UBGABgG4tcAnASwHMALE8igBwA8bMBfLCAA0IQgwIA5JAFs4iGN1mcCAOjABnDSPARM4pghAZMKFOhCKcMTFA0XSAbSxmzFjTgCeBPA0w4AWkh9ZgCPJFskOigLFxQAXWE4izoAVwJ7BDQ4txBYMDwopBwGPQDOOghOGDovAIA3JDxUmADMCADUzABrdoB3bCycNJg4wQEdYIMAMQg6GWKjACsNPR1YTm1EE1yPbxhffwdzEELrDwsk0xO9nz9AqdDwyOjjizOYC5AsfhB+IA)

### What minimal configuration is needed to reproduce the bug?

```json
{
  "extends": [
    "stylelint-config-standard"
  ],
  "rules": {
    "declaration-property-value-no-unknown": true
  }
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgGxgDMd4UBGABgG4tcAnASwHMALE8igBwA8bMBfLCAA0IQgwIA5JAFs4iGN1mcCAOjABnDSPARM4pghAZMKFOhCKcMTFA0XSAbSxmzFjTgCeBPA0w4AWkh9ZgCPJFskOigLFxQAXWE4izoAVwJ7BDQ4txBYMDwopBwGPQDOOghOGDovAIA3JDxUmADMCADUzABrdoB3bCycNJg4wQEdYIMAMQg6GWKjACsNPR1YTm1EE1yPbxhffwdzEELrDwsk0xO9nz9AqdDwyOjjizOYC5AsfhB+IA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "latest",
  "stylelint-config-standard": "latest"
}
```

### What did you expect to happen?

I would expect `declaration-property-value-no-unknown` to **specifically** only complain about properties that are invalid. Also, some browsers will interpret `10` as `10px`. So `left: 10;` actually works.

### What actually happened?

```
Unexpected unknown value "10" for property "left"
[(declaration-property-value-no-unknown)](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown)
```

### Do you have a proposal to fix the bug?

I propose introducing a new rule, `length-unit`, that requires (where possible) or disallows units. This rule would clearly distinguish between truly invalid properties and those where a unit is missing. By implementing this, we can reduce false positives and provide clearer guidance to users, making it easier to fix linting issues.

**Comment by jeddy3** [comment]
@pamelalozano16 Thanks for the report and for using the template.

> Also, some browsers will interpret 10 as 10px. So left: 10; actually works.

It'd be great to get more details about this, particularly which browsers do and don't support this syntax.

As the specifications say: 

- [`left` accepts `auto |<length-percentage>`](https://drafts.csswg.org/css-position/#insets)
- [`<length>` is a `<dimension>`](https://drafts.csswg.org/css-values-4/#lengths)
- [a `<dimension>` is a "number with a unit attached to it"](https://drafts.csswg.org/css-values-4/#dimensions)

But:

> For zero lengths the unit identifier is optional (i.e. can be syntactically represented as the `<number>`).

For which we have the [`length-zero-no-unit`](https://stylelint.io/user-guide/rules/length-zero-no-unit/) rule.

The [`declaration-property-value-no-unknown` rule](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown/) is there to help people write spec-compliant CSS:

> This rule considers values for properties defined within the CSS specifications to be known. You can use the `propertiesSyntax` and `typesSyntax` secondary options to extend the syntax.

You can extend the syntax to allow numbers using these secondary options, e.g. by adding a new `<length-number>` type:

```json
{
  "extends": [
    "stylelint-config-standard"
  ],
  "rules": {
    "declaration-property-value-no-unknown": [true, {
      "propertiesSyntax": {
        "left": "| <length-number>"
      },
      "typesSyntax": { "length-number": "| <number>" }
    }]
  }
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgGxgDMd4UBGABgG4tcAnASwHMALE8igBwA8bMBfLCAA0IQgwIA5JAFs4iGN1mcCAOjABnDSPARM4pghAZMKFOhCKcMTFA0XSAbSxmzFjTgCeBPA0w4AWkh9ZgCPJFskOigLFxQAXWE4izoAVwJ7BDQ4txBYMDwopBwGPQDOOghOGDovAIA3JDxUmADMCADUzABrdoB3bCzHHDSYYWzTV1yKqpqSmA0AZU9-JG4HCampiwJiDYsAHxQAHgJMJhwWNtSZACMagD5Yyan+JJfXCy9qpZWcNY2wHMIDOFyumBu9zo+xAR2OELujwsKEEH348TiqP4OmCBgAYhA6DJikYAFYaPQ6WCcbSIEy5DzeGC+fwwwrWDwWd4Mrw+PyBXGhcKRaJs4oLHDPbH8IA).

**Comment by nex3** [comment]
@jeddy3 I think you're missing [this part of the spec for `<length>`](https://drafts.csswg.org/css-values-4/#lengths):

> For zero lengths the unit identifier is optional (i.e. can be syntactically represented as the [`<number>`](https://drafts.csswg.org/css-values-4/#number-value) 0).

`left: 0` is definitely legal per spec, which is why it's universally supported by browsers.

**Comment by jeddy3** [comment]
Thanks for chiming in, @nex3. 

I mentioned that part of the spec, along with the `length-zero-no-unit` rule, towards the middle of my post as the `declaration-property-value-no-unknown` rule doesn't flag:

```css
a {
  left: 0; /* not flagged */
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgGxgDMd4UAGAbiwF8sQAaEQgSwIDkkBbORGADy4AHAgDowAZ3ENwETCwDmCEBkwoU6EPxwxMUcRtIBtLGrUbxOAJ4E8zTDgC0kOc3kOLSXUgBOUDSZQAXXoAjW8AVwJ9BDQAsxBYMDwfJBxmWQdBbwhBGG8rBwA3JDxwmAdMCAdwzABrSoB3bBicCJgA2kxqaWcFADEIb05UpQArcVlpWEEpRBV4i2sYW3sDdRBk7QsNENV1xZs7R17XdxxPKB8-GI1NmG2QGhBqIA)

Only non-zero numbers, with `10` as the example in the [original post](https://github.com/stylelint/stylelint/issues/8096#issue-2621920355), are flagged by the rule:

```css
a {
  left: 10; /* flagged */
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgGxgDMd4UBGABgG4sBfLEAGhEIEsCA5JAWzkRgAePAA4EAdGADOkpuAiY2AcwQgMmFCnQhBOGJiiStpANpYNGrZJwBPAnlaYcAWkgLWip1aT6kAJyhaZigAuoxBWr4ArgSGCGhBFiCwYHh+SDis8k7CvhDCML42TgBuSHiRME6YEE6RmADW1QDu2HE4UTBB9Ji0sq5KAGIQvtzpKgBWkvKysMIyiGqJVrYw9o5GmiCpulZaYeqby3YOzv3unjjeUH4BcVrbMLsgdCC0QA)

It'd be great to know more about which browsers support non-zero numbers for `<length>` so we can decide if the rule behaves in a way that's useful to people.

**Comment by nex3** [comment]
Oh, you're right, I missed the context that this was for actual numbers. I'm checking with Pamela and internal users to see if this is a real issue.

**Comment by Mouvedia** [comment]
Ill give you an analogy for something that is specced.

> [An `image` start tag token is handled by the tree builder, but it is not in this list because it is not an element; it gets turned into an `img` element.](https://html.spec.whatwg.org/multipage/parsing.html#the-stack-of-open-elements:the-img-element-2)

Still we didn't add `'image'` to

https://github.com/stylelint/stylelint/blob/42cd4757369706178fa2600dcca054b068682bd9/lib/reference/selectors.mjs#L5

---

I think this issue should be closed because the unitless fallback behaviour to `px` is not a trick which is worth advertising even if it happened to be widely supported by existing browsers.

**Comment by nex3** [comment]
It looks like the reports we were seeing about this were unexpectedly using quirks mode, so I'm going to close this out.

**Comment by Mouvedia** [comment]
ref: https://drafts.csswg.org/css-values-4/#deprecated-quirky-length
